File: common/src/main/java/com/mcal/common/utils/ScopedStorage.java
package com.mcal.common.utils;

import android.os.Environment;

import androidx.annotation.NonNull;

import com.mcal.common.App;

import org.jetbrains.annotations.Contract;

import java.io.File;
import java.util.List;

public class ScopedStorage {
    @NonNull
    @Contract(" -> new")
    public static File getApkEditorDirectory() {
        return new File(Environment.getExternalStorageDirectory() + "/ApkEditor");
    }

    public static File getFilesDir() {
        return App.getContext().getFilesDir();
    }

    public static String getExternalStoragePath() {
        String internalPath = ScopedStorage.getStorageDirectory().getPath();
        File[] files = App.getContext().getExternalFilesDirs(null);
        if (files != null) {
            // Find the pattern
            int appendedLen = 0;
            for (File f : files) {
                String path = f.getPath();
                if (path.startsWith(internalPath)) {
                    appendedLen = path.length() - internalPath.length();
                    break;
                }
            }
            for (File f : files) {
                String path = f.getPath();
                if (!path.startsWith(internalPath)) {
                    return path.substring(0, path.length() - appendedLen);
                }
            }
        }

        String path = null;
        List<StorageUtils.StorageInfo> storageList = StorageUtils.getStorageList();
        if (storageList != null) {
            for (StorageUtils.StorageInfo si : storageList) {
                if (!si.internal && !si.readonly) {
                    path = si.path;
                    break;
                }
            }
        }
        return path;
    }

    public static File getStorageDirectory() {
        return Environment.getExternalStorageDirectory();
    }
}

File: common/src/main/java/com/mcal/common/utils/ScopedStorage.kt
package com.mcal.common.utils

import android.os.Environment
import com.mcal.common.App
import java.io.File

object ScopedStorage {
    @JvmStatic
    val filesDir: File
        get() = App.getContext().filesDir

    @JvmStatic
    val cacheDir: File
        get() = App.getContext().cacheDir

    @JvmStatic
    val storageDirectory: File
        get() = Environment.getExternalStorageDirectory()

    @JvmStatic
    fun getDecodedDir(): File {
        val path = File(filesDir.path + File.separator + "decoded")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getApkEditorDir(): File {
        val path = File("$storageDirectory/ApkEditor")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getBinDir(): File {
        val path = File(filesDir.path + File.separator + "bin")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getTmpDir(): File {
        val path = File(getApkEditorDir().path + File.separator + "tmp")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getTempDir(): File {
        val path = File(getApkEditorDir().path + File.separator + "temp")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getBackupsDir(): File {
        val path = File(getApkEditorDir().path + File.separator + "backups")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getDataDir(): File {
        val path = File(filesDir.path + File.separator + "data")
        if (!path.exists()) path.mkdirs()
        return path
    }

    @JvmStatic
    fun getPublicKey(): File {
        return File(getBinDir(), "testkey.x509.pem")
    }

    @JvmStatic
    fun getPrivateKey(): File {
        return File(getBinDir(), "testkey.pk8")
    }

    @JvmStatic
    fun getAapt(): File {
        val path = File(getBinDir(), "aapt")
        path.setExecutable(true)
        return path
    }

    @JvmStatic
    fun getAapt2(): File {
        val path = File(getBinDir(), "aapt2")
        path.setExecutable(true)
        return path
    }
}